### PR TITLE
feat(ios): added method and properties for native tag technology class

### PR DESCRIPTION
**Pull request by anil-shukla-axway**
https://jira.appcelerator.org/browse/MOD-2834

### CODE REVIEW

**Review by vikas-goyal-axway — CHANGES_REQUESTED** [review]

**Comment by anil-shukla-axway** [comment]
@satinder-singh-axway  @vikas-goyal-axway  @fasihul-axway  Please review this PR.

File: ios/Classes/TagTechnologyProxy.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface TagTechnology : TiProxy
+
+- (void)connect:(id)unused;
+
+- (void)close:(id)unused;
+
+- (void)restartPolling:(id)unused;
```
Comment (vikas-goyal-axway) [inline ios/Classes/TagTechnologyProxy.h:19]: remove it if this function is not required

File: ios/Classes/TagTechnologyProxy.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#import "TiProxy.h"
+#import <CoreNFC/CoreNFC.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface TagTechnology : TiProxy
+
+- (void)connect:(id)unused;
+
+- (void)close:(id)unused;
```
Comment (vikas-goyal-axway) [inline ios/Classes/TagTechnologyProxy.h:17]: remove it if this function is not required

File: ios/Classes/TagTechnologyProxy.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+- (void)restartPolling:(id)unused;
+
+- (NSNumber *)isEnabled:(id)unused;
```
Comment (vikas-goyal-axway) [inline ios/Classes/TagTechnologyProxy.h:21]: remove it if this function is not required

File: ios/Classes/TagTechnologyProxy.m [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+- (void)connect:(id)unused
+{
+}
+
+- (void)close:(id)unused
+{
+}
+
+- (void)restartPolling:(id)unused
```
Comment (vikas-goyal-axway) [inline ios/Classes/TagTechnologyProxy.m:25]: remove it if this function is not required

File: ios/Classes/TagTechnologyProxy.m [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+@implementation TagTechnology
+
+#pragma mark Public API's
+
+- (void)connect:(id)unused
+{
+}
+
+- (void)close:(id)unused
```
Comment (vikas-goyal-axway) [inline ios/Classes/TagTechnologyProxy.m:21]: remove it if this function is not required

File: ios/Classes/TagTechnologyProxy.m [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+- (void)restartPolling:(id)unused
+{
+}
+
+- (NSNumber *)isConnected:(id)unused
+{
+  return;
+}
+
+- (NSNumber *)isEnabled:(id)unused
```
Comment (vikas-goyal-axway) [inline ios/Classes/TagTechnologyProxy.m:34]: remove it if this function is not required

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+          [self fireEvent:@"didDetectTags"
+               withObject:@{
+                 @"errorCode" : NUMINTEGER([error code]),
+                 @"errorDescription" : [error localizedDescription],
+               }];
+        } else {
+          [self fireEvent:@"didDetectTags"
+               withObject:@{
+                 @"tag" : tag,
+                 @"errorCode" : NUMINTEGER([error code]),
```
Comment (vikas-goyal-axway) [inline ios/Classes/NativeTagTechnologyProxy.m:31]: error object would be nil here so why we are using nil object

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                 @"errorDescription" : [error localizedDescription],
```
Comment (vikas-goyal-axway) [inline ios/Classes/NativeTagTechnologyProxy.m:32]: same here , error object would be nil here so why we are using nil object

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        if (error != nil) {
+          [self fireEvent:@"didDetectTags"
+               withObject:@{
+                 @"errorCode" : NUMINTEGER([error code]),
+                 @"errorDescription" : [error localizedDescription],
+               }];
+        } else {
+          [self fireEvent:@"didDetectTags"
+               withObject:@{
+                 @"tag" : tag,
```
Comment (vikas-goyal-axway) [inline ios/Classes/NativeTagTechnologyProxy.m:30]: do not send native tag object, you should send TagProxy class here

**Review by vikas-goyal-axway — CHANGES_REQUESTED** [review]

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                 @"errorCode" : NUMINTEGER([error code]),
+                 @"errorDescription" : [error localizedDescription],
+               }];
+        }
+      }];
+}
+
+- (NSNumber *)isConnected:(id)unused
+{
+  if (session.connectedTag != nil) {
```
[the remark this answers is not in the document]
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m]: Updated the code

File: ios/Classes/TagTechnologyProxy.h (hunk 2 above)
Comment (anil-shukla-axway) [reply ios/Classes/TagTechnologyProxy.h]: removed

File: ios/Classes/TagTechnologyProxy.h (hunk 1 above)
Comment (anil-shukla-axway) [reply ios/Classes/TagTechnologyProxy.h]: removed

File: ios/Classes/TagTechnologyProxy.h (hunk 3 above)
Comment (anil-shukla-axway) [reply ios/Classes/TagTechnologyProxy.h]: removed

File: ios/Classes/TagTechnologyProxy.m (hunk 4 above)
Comment (anil-shukla-axway) [reply ios/Classes/TagTechnologyProxy.m]: removed

File: ios/Classes/TagTechnologyProxy.m (hunk 5 above)
Comment (anil-shukla-axway) [reply ios/Classes/TagTechnologyProxy.m]: removed

File: ios/Classes/TagTechnologyProxy.m (hunk 6 above)
Comment (anil-shukla-axway) [reply ios/Classes/TagTechnologyProxy.m]: removed

File: ios/Classes/NativeTagTechnologyProxy.m (hunk 7 above)
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m]: Updated the code

File: ios/Classes/NativeTagTechnologyProxy.m (hunk 8 above)
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m]: Updated the code

File: ios/Classes/NativeTagTechnologyProxy.m (hunk 9 above)
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m]: Created proxy for NFC tag and send object for same

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+@implementation NativeTagTechnology
+
+#pragma mark Public API's
+
+- (void)connect:(id)unused
+{
+  [session connectToTag:tag
+      completionHandler:^(NSError *_Nullable error) {
+        if (error != nil) {
+          [self fireEvent:@"didDetectTags"
```
Comment (vikas-goyal-axway) [inline ios/Classes/NativeTagTechnologyProxy.m:23]: @anil-shukla-axway , I think it is connect event so you should call connect tag event here

**Review by vikas-goyal-axway — CHANGES_REQUESTED** [review]
@anil-shukla-axway , please add YML for NFCTag as well

File: apidoc/NfcTag.yml [hunk 12]
```diff
@@ -6,7 +6,7 @@ description: |
     [Tag](http://developer.android.com/reference/android/nfc/Tag.html)
 extends: Titanium.Proxy
 since: "1.0.0"
-platforms: [android]
+platforms: [android, iphone]
```
Comment (vikas-goyal-axway) [inline apidoc/NfcTag.yml:9]:
```suggestion
platforms: [android, iphone, ipad]
```

File: ios/Classes/NativeTagTechnologyProxy.m (hunk 11 above)
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m:23]: didDetectTags events is the events where we connect with the tag. after beginning the session, when any tag got found, the didDetectTags event get called.

Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m]: Updated the code.

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+@implementation NativeTagTechnology
+
+#pragma mark Public API's
+
+- (void)connect:(id)unused
+{
+  [session connectToTag:tag
+      completionHandler:^(NSError *_Nullable error) {
+        [self fireEvent:@"didConnectTag"
```
Comment (vikas-goyal-axway) [inline ios/Classes/NativeTagTechnologyProxy.m:22]: I think before calling event, you should also check if event is set or not.
```
if (![self _hasListeners:@"didConnectTag"]) {
    return;
  }
```

File: ios/Classes/NativeTagTechnologyProxy.m [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+             }];
+      }];
+}
+
+- (NSNumber *)isConnected:(id)unused
+{
+  if (session.connectedTag == tag) {
+    return [NSNumber numberWithBool:YES];
+  } else {
+    return [NSNumber numberWithBool:NO];
```
Comment (fasihul-axway) [inline ios/Classes/NativeTagTechnologyProxy.m:37]: I think the whole condition can be simplified like this 

`return [NSNumber numberWithBool:session.connectedTag == tag?YES:NO];`

**Review by fasihul-axway — COMMENTED** [review]
I think class name should start with prefix **TiNfc**, also methods of class **TiNFCTagProxy** is not documented. 

**Note:** Add method and event to [wiki](https://wiki.appcelerator.org/display/~ssingh/Native+TagTechnology) as well.

File: ios/Classes/NativeTagTechnologyProxy.m (hunk 13 above)
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m:22]: Updated accordingly.

File: ios/Classes/NativeTagTechnologyProxy.m (hunk 14 above)
Comment (anil-shukla-axway) [reply ios/Classes/NativeTagTechnologyProxy.m:37]: Updated code.

**Review by vikas-goyal-axway — APPROVED** [review]

**Review by fasihul-axway — APPROVED** [review]

**Pull request merged**